### PR TITLE
Use #select instead of guard

### PR DESCRIPTION
**Pull request by yitznewton**

### CODE REVIEW

File: lib/shape/data_visitor.rb [hunk 1]
```diff
@@ -8,13 +8,15 @@ def visit(visitor = lambda {|x| x})
...
+      all_properties.select { |_, property| visitable?(property) }
+    end
+
+    def all_properties
       self.class.properties.merge(self.properties)
     end
 
     def data_visitor(properties = self.properties_to_visit, visitor = lambda {|x| x})
       properties.each_with_object({}) do |(name, property), obj|
-        next unless visitable?(property)
```
Comment (yitznewton) [inline lib/shape/data_visitor.rb]: Hmm, this means that when passing in properties manually, the filter won't be applied

Comment (yitznewton) [inline lib/shape/data_visitor.rb]: Up to you whether you want the onus to be on the calling code or not

Comment (robincurry) [inline lib/shape/data_visitor.rb]: It seems like putting the onus on the caller could result in surprises. Are there benefits to using the select over guard that would make that tradeoff worthwhile?

Comment (yitznewton) [inline lib/shape/data_visitor.rb]: Just FP sweetness, and an even higher level of declarativeness/confident-code-ness. i.e.

> visit all the properties I'm supposed to visit (which are already filtered)

as opposed to

> visit all the properties I'm supposed to visit... oh, but only if visitable

Do we need to provide those defaults on #data_visitor?

Comment (yitznewton) [inline lib/shape/data_visitor.rb]: #data_visitor is protected, and I don't see anywhere we're calling it directly, at least not in platform-api

**Comment by yitznewton** [comment]
... so we could even do (that second commit)

File: lib/shape/data_visitor.rb [hunk 2]
```diff
@@ -8,13 +8,15 @@ def visit(visitor = lambda {|x| x})
...
       self.class.properties.merge(self.properties)
     end
 
-    def data_visitor(properties = self.properties_to_visit, visitor = lambda {|x| x})
+    def data_visitor(properties, visitor)
       properties.each_with_object({}) do |(name, property), obj|
-        next unless visitable?(property)
-
         if property.properties.present?
           obj[name] = self.data_visitor(property.properties, visitor)
```
Comment (robincurry) [inline lib/shape/data_visitor.rb]: I think the problem with the select (as currently implemented) would show up right here :point_up: - essentially, only top-level visitables would work. We aren't ensuring that visitable is called as we recurse nested properties.

Comment (yitznewton) [inline lib/shape/data_visitor.rb]: Ah, I missed the recursion there.

**Pull request closed without merging**